packages/kotti-ui/source/props.test.ts

```diff
@@ -0,0 +1,451 @@
+}
+
+const expectValidator = (
+	prop: PropOptions<unknown, boolean>,
+	successCases: unknown[],
+	failCases: unknown[],
+) => {
+	const { validator } = prop
+
+	if (validator === undefined) {
+		expect(validator).not.toBe(undefined)
+		throw new Error('unexpected undefined validator')
+	}
+
+	successCases.forEach((c) => expect(validator(c)).toBe(true))
+	failCases.forEach((c) => expect(validator(c)).toBe(false))
+}
+
+describe('array', () => {
+	beforeAll(silenceConsole)
+
+	it('z.array(z.number())', () => {
+		const schema = z.object({
+			prop: z.array(z.number()),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Array)
+		expectValidator(prop, [[], [1], [-1]], [1, null, 'test', undefined])
+	})
+
+	it('z.array(z.number()).nullable()', () => {
+		const schema = z.object({
+			prop: z.array(z.number()).nullable(),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Array)
+		expectValidator(
+			prop,
+			[[], [1], [-1], null],
+			[[null], ['string'], 1, 'test', undefined],
+		)
+	})
+
+	it('z.array(z.number()).default()', () => {
+		const schema = z.object({
+			prop: z.array(z.number()).default(() => []),
+		})
+
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, [])
+		expect(prop.type).toBe(Array)
+		expectValidator(
+			prop,
+			[[], [1], [-1], undefined],
+			[[null], ['string'], 1, null, 'test'],
+		)
+	})
+
+	it('z.array(z.number()).nullable().default()', () => {
+		const schema = z.object({
+			prop: z.array(z.number()).nullable().default(null),
+		})
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, null)
+		expect(prop.type).toBe(Array)
+		expectValidator(
+			prop,
+			[[], [1], [-1], null, undefined],
+			[[null], ['string'], 1, 'test'],
+		)
+	})
+})
+
+describe('boolean', () => {
+	beforeAll(silenceConsole)
+
+	it('z.boolean()', () => {
+		const schema = z.object({
+			prop: z.boolean(),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Boolean)
+		expectValidator(prop, [true, false], [null, undefined])
+	})
+
+	it('z.boolean().nullable()', () => {
+		const schema = z.object({
+			prop: z.boolean().nullable(),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Boolean)
+		expectValidator(prop, [true, false, null], ['true', undefined])
+	})
+
+	it('z.boolean().default()', () => {
+		const schema = z.object({
+			prop: z.boolean().default(true),
+		})
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, true)
+		expect(prop.type).toBe(Boolean)
+		expectValidator(prop, [true, false, undefined], [null])
+	})
+
+	it('z.boolean().nullable().default()', () => {
+		const schema = z.object({
+			prop: z.boolean().nullable().default(null),
+		})
+		const { prop } = makeProps(schema)
+
+		expectDefault(prop, null)
+		expect(prop.type).toBe(Boolean)
+		expectValidator(prop, [true, false, null, undefined], ['true'])
```

Comment:
the fail scenario is a string true for a fail scenario, unrelated to `nullable()` and `default()` 

up to you to keep it on this test or move it out